File: labs-java/src/main/java/labs/Lab07.java
package labs;

import java.util.ArrayList;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.azure.cosmos.CosmosAsyncContainer;
import com.azure.cosmos.CosmosAsyncDatabase;
import com.azure.cosmos.models.CosmosStoredProcedureRequestOptions;
import com.azure.cosmos.models.PartitionKey;
import com.fasterxml.jackson.databind.ObjectMapper;

import labs.Models.DeleteStatus;
import labs.Models.Food;
import reactor.core.publisher.Mono;
import services.CosmosDBService;
import services.DataFaker;

public class Lab07 {
	private static final Logger logger = LoggerFactory.getLogger(Lab07.class);
	private CosmosDBService cosmosDBService;
	private static ObjectMapper mapper = new ObjectMapper();
	private static int pointer = 0;
	private static boolean resume = false;

	/**
	 * Constructor
	 * 
	 * @param cosmosDBService The Cosmos DB service instance
	 * @throws Exception All possible exceptions that can be thrown
	 */
	public Lab07(final CosmosDBService cosmosDBService) throws Exception {
		this.cosmosDBService = cosmosDBService;
		String databaseId = "NutritionDatabase";
		String containerId = "FoodCollection";

		// Get database and container
		CosmosAsyncDatabase targetDatabase = this.cosmosDBService.getDatabase(databaseId);
		CosmosAsyncContainer targetContainer = this.cosmosDBService.getContainer(targetDatabase, containerId);

		// Bulk Upload
		executeBulkUpload(targetContainer);

		// Bulk Delete
		executeBulkDelete(targetContainer);
	}

	/**
	 * Execute Bulk upload.
	 * 
	 * @param container The container to upload the items to
	 */
	private void executeBulkUpload(final CosmosAsyncContainer container) {
		DataFaker dataFaker = new DataFaker();
		List<Food> foods = dataFaker.generateFood(1000);

		String sprocName = "bulkUpload";
		CosmosStoredProcedureRequestOptions options = new CosmosStoredProcedureRequestOptions();
		options.setPartitionKey(new PartitionKey("Energy Bars"));

		while (pointer < foods.size()) {
			List<Object> sprocArgs = new ArrayList<Object>();
			List<Food> foodObjectsToUpload = foods.subList(pointer, foods.size());
			sprocArgs.add(foodObjectsToUpload);

			this.cosmosDBService.executeStoredProcedureAsync(container, sprocName, options, sprocArgs)
					.flatMap(response -> {
						int delta_items = Integer.parseInt(response.getResponseAsString());
						pointer += delta_items;
						logger.info("{} Total Items {} Items Uploaded in this Iteration...", pointer, delta_items);
						return Mono.empty();
					}).block();
		}
	}

	/**
	 * Execute Bulk delete.
	 * 
	 * @param container The container where to delete the items
	 */
	private void executeBulkDelete(final CosmosAsyncContainer container) {
		resume = true;
		String sprocName = "bulkDelete";
		String query = "SELECT * FROM foods f WHERE f.foodGroup = 'Energy Bars'";

		CosmosStoredProcedureRequestOptions options = new CosmosStoredProcedureRequestOptions();
		options.setPartitionKey(new PartitionKey("Energy Bars"));

		List<Object> sprocArgs = new ArrayList<Object>();
		sprocArgs.add(query);

		do {
			this.cosmosDBService.executeStoredProcedureAsync(container, sprocName, options, sprocArgs)
					.flatMap(executeResponse -> {

						DeleteStatus result = null;

						try {
							result = mapper.readValue(executeResponse.getResponseAsString(), DeleteStatus.class);
						} catch (Exception ex) {
							logger.error("Failed to parse bulkDelete response.", ex);
						}

						logger.info("Batch Delete Completed. Deleted: {} Continue: {} Items Uploaded in this Iteration",
								result.getDeleted(), result.isContinuation());

						resume = result.isContinuation();

						return Mono.empty();
					}).block();
		} while (resume);
	}
}

File: sql-api/csharp/v3/Solution/Labs/Lab07.cs
using System;
using System.Linq;
using System.Threading.Tasks;
using System.Threading;
using System.Collections.Generic;
using Microsoft.Azure.Cosmos;
using Microsoft.Azure.Cosmos.Scripts;
using Solution.Entities;
using Solution.Entities.Lab07;


namespace Solution.Labs
{
    public class Lab07
    {
        private readonly string StoredProcedureBulkUpload = @"
function bulkUpload(docs) {
    var container = getContext().getCollection();
    var containerLink = container.getSelfLink();
    var count = 0;
    if (!docs) throw new Error('The array is undefined or null.');
    var docsLength = docs.length;
    if (docsLength == 0) {
        getContext()
            .getResponse()
            .setBody(0);
        return;
    }
    tryCreate(docs[count], callback);
    function tryCreate(doc, callback) {
        var isAccepted = container.createDocument(containerLink, doc, callback);
        if (!isAccepted)
            getContext()
                .getResponse()
                .setBody(count);
    }
    function callback(err, doc, options) {
        if (err) throw err;
        count++;
        if (count >= docsLength) {
            getContext()
                .getResponse()
                .setBody(count);
        } else {
            tryCreate(docs[count], callback);
        }
    }
}
        ";

        private readonly string StoredProcedureBulkDelete = @"
function bulkDelete(query) {
    var container = getContext().getCollection();
    var containerLink = container.getSelfLink();
    var response = getContext().getResponse();
    var responseBody = {
        deleted: 0,
        continuation: true
    };
    if (!query) throw new Error('The query is undefined or null.');
    tryQueryAndDelete();
    function tryQueryAndDelete(continuation) {
        var requestOptions = { continuation: continuation };
        var isAccepted = container.queryDocuments(
            containerLink,
            query,
            requestOptions,
            function (err, retrievedDocs, responseOptions) {
                if (err) throw err;
                if (retrievedDocs.length > 0) {
                    tryDelete(retrievedDocs);
                } else if (responseOptions.continuation) {
                    tryQueryAndDelete(responseOptions.continuation);
                } else {
                    responseBody.continuation = false;
                    response.setBody(responseBody);
                }
            }
        );
        if (!isAccepted) {
            response.setBody(responseBody);
        }
    }
    function tryDelete(documents) {
        if (documents.length > 0) {
            var isAccepted = container.deleteDocument(
                documents[0]._self,
                {},
                function (err, responseOptions) {
                    if (err) throw err;
                    responseBody.deleted++;
                    documents.shift();
                    tryDelete(documents);
                }
            );
            if (!isAccepted) {
                response.setBody(responseBody);
            }
        } else {
            tryQueryAndDelete();
        }
    }
}
        ";

        /// <summary>
        ///     Constructor
        /// </summary>
        /// <param name="configuration">
        ///     Configuration object
        /// </param>
        public Lab07(IConfiguration _configuration)
        {
            string cosmosDBURI = _configuration.CosmosDB.URI;
            string cosmosDBPrimaryKey = _configuration.CosmosDB.PrimaryKey;
            string cosmosDBConnectionString = _configuration.CosmosDB.PrimaryConnectionString;
            CosmosClient cosmosDBClient = new CosmosClient(cosmosDBURI, cosmosDBPrimaryKey);

            string databaseId = "NutritionDatabase";
            Database nutritionDatabase = cosmosDBClient.GetDatabase(databaseId);
            string containerId = "FoodCollection";
            Container foodContainer = cosmosDBClient.GetContainer(databaseId, containerId);

            // Register Stored Procedures
            UpsertStoredProcedure(foodContainer, "bulkUpload", StoredProcedureBulkUpload).GetAwaiter().GetResult();
            UpsertStoredProcedure(foodContainer, "bulkDelete", StoredProcedureBulkDelete).GetAwaiter().GetResult();
            Thread.Sleep(10000);

            // Execute Stored Procedures
            BulkUpload(foodContainer).GetAwaiter().GetResult();
            BulkDelete(foodContainer).GetAwaiter().GetResult();
        }

        /// <summary>
        ///     Upsert a Stored Procedure
        /// </summary>
        /// <param name="container">
        ///     Container Reference
        /// </param>
        /// <param name="storedProcedureId">
        ///     The id of the Stored Procedure
        /// </param>
        /// <param name="storedProcedureBody">
        ///     The body of the Stored Procedure
        /// </param>
        /// <returns>
        ///     Returns a task
        /// </returns>
        private async Task UpsertStoredProcedure(
            Container container,
            string storedProcedureId,
            string storedProcedureBody)
        {
            Console.WriteLine($"Processing Stored Procedure '{storedProcedureId}'...");

            StoredProcedureProperties bulkUploadProperties = new StoredProcedureProperties
            {
                Id = storedProcedureId,
                Body = storedProcedureBody
            };

            try
            {
                StoredProcedureResponse storedProcedureResponse = await container.Scripts.CreateStoredProcedureAsync(bulkUploadProperties);
                Console.WriteLine($"Stored Procedure Response Status Code for creation: {storedProcedureResponse.StatusCode}...");
            }
            catch
            {
                StoredProcedureResponse storedProcedureResponse = await container.Scripts.ReplaceStoredProcedureAsync(bulkUploadProperties);
                Console.WriteLine($"Stored Procedure Response Status Code for replacement: {storedProcedureResponse.StatusCode}...");
            }
        }

        /// <summary>
        ///     Bulk upload documents
        /// </summary>
        /// <param name="container">
        ///     Container reference
        /// </param>
        /// <returns>
        ///     Returns a task
        /// </returns>
        private static async Task BulkUpload(Container container)
        {
            List<Food> foods = new Bogus.Faker<Food>()
                .RuleFor(p => p.Id, f => (-1 - f.IndexGlobal).ToString())
                .RuleFor(p => p.Description, f => f.Commerce.ProductName())
                .RuleFor(p => p.ManufacturerName, f => f.Company.CompanyName())
                .RuleFor(p => p.FoodGroup, f => "Energy Bars")
                .Generate(10000);

            int pointer = 0;

            while (pointer < foods.Count)
            {
                StoredProcedureExecuteResponse<int> result = await container.Scripts.ExecuteStoredProcedureAsync<int>(
                    "bulkUpload",
                    new PartitionKey("Energy Bars"),
                    new dynamic[] { foods.Skip(pointer).Take(1000) });

                pointer += result.Resource;

                await Console.Out.WriteLineAsync($"{pointer} Total Items\t{result.Resource} Items Uploaded in this Iteration...");
            }
        }

        /// <summary>
        ///     Bulk delete documents
        /// </summary>
        /// <param name="container">
        ///     Container reference
        /// </param>
        /// <returns>
        ///     Returns a task
        /// </returns>
        private static async Task BulkDelete(Container container)
        {
            bool resume = true;
            do
            {
                string query = "SELECT * FROM foods f WHERE f.foodGroup = 'Energy Bars'";
                StoredProcedureExecuteResponse<DeleteStatus> result = await container.Scripts.ExecuteStoredProcedureAsync<DeleteStatus>("bulkDelete", new PartitionKey("Energy Bars"), new dynamic[] { query });

                await Console.Out.WriteLineAsync($"Batch Delete Completed.\tDeleted: {result.Resource.Deleted}\tContinue: {result.Resource.Continuation}");
                resume = result.Resource.Continuation;
            }
            while (resume);
        }
    }
}